File: prob0023/NonAbundant.java
package prob0023;

import java.lang.Math;
import java.util.Arrays;
public class NonAbundant {
	protected static int[] copyArray(int[] in, int newSize) {
		int[] out = new int[newSize];
		System.arraycopy(in, 0, out, 0, (newSize > in.length ? in.length : newSize));
		return out;
	}
	public static int[] properDivisors(int num) {
		int root = (int)Math.sqrt((double)num);
		int[] out = {1};
		for (int i = 2; i <= root; i++) {
			if (num % i == 0) {
				if (i * i == num) {
					out = copyArray(out, out.length + 1);
					out[out.length-1] = i;
				}
				else {
					out = copyArray(out, out.length + 2);
					out[out.length-1] = i;
					out[out.length-2] = num / i;
				}
			}
		}
		Arrays.sort(out);
		return out;
	}
	protected static void printarr(int[] in) {
		System.out.print("[ " + in[0]);
		for (int i = 1; i < in.length; i++)
			System.out.print(" , " + in[i]);
		System.out.println(" ]");
	}

	public static void main(String[] args) {
		int[] abundants = new int[0];
		int nextAbundant = 0;
		//Build list of abundants
		for (int i = 2; i <= 28123; i++) {
			int[] divisors = properDivisors(i);
			int sum = 0;
			for (int div : divisors)
				sum += div;
			if (sum > i) {
				if (nextAbundant == abundants.length)
					abundants = copyArray(abundants, abundants.length + 100);
				abundants[nextAbundant++] = i;
				//System.out.print(i + " "); printarr(divisors);
			}
		}

		//Determine which vales are sums of
		boolean[] isSum = new boolean[28124];
		for (int i = 0; i < nextAbundant; i++) {
			for (int j = i; abundants[i] + abundants[j] <= 28123; j++) {
				isSum[abundants[i] + abundants[j]] = true;
			}
		}

		//Total those not representable as a sum
		int total = 0;
		for (int i = 1; i < 28124; i++)
			if (!(isSum[i]))
				total += i;
		System.out.println("Total of positive integers that cannot be represented" +
			" as a sum of abundant numbers: " + total);
	}
}

File: prob0023/nonabundant.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <string.h>
#include <gmp.h>

//Returns a zero-terminated int array of divisors
int *divisors(int const num) {
	int len = 3, *list = malloc(len * sizeof(int));
	int const root = (int)sqrt((double)num);
	list[0] = 1; list[1] = num;
	for (int i = 2; i <= root; i++) {
		if (num % i == 0) {
			if (num / i != i) { //Two different factors
				len += 2;
				list = realloc(list, len * sizeof(int));
				if (list == NULL) {
					fprintf(stderr, "Error allocating memory\n");
					exit(1);
				}
				list[len-3] = i;
				list[len-2] = num / i;
			}
			else { // i * i == num
				len += 1;
				list = realloc(list, len * sizeof(int));
				if (list == NULL) {
					fprintf(stderr, "Error allocating memory\n");
					exit(1);
				}
				list[len-2] = i;
			}
		}
	}
	list[len-1] = 0;
	return list;
}
int arrsum(int const *arr) {
	int sum = 0;
	for (int i = 0; arr[i]; i++) {
		sum += arr[i];
	}
	return sum;
}

int main(int argc, char const *argv[]) {
	int abundants[8192], len = 0;
	long unsigned total = 0;
	char absums[28213];
	memset(&(abundants[0]), 0, sizeof(int)*8192);
	memset(&(absums[0]), 1, sizeof(char)*28213);
	
	//Determine which are abundant
	for (int i = 2; i < 28123; i++) {
		int *divs = divisors(i), divsum = arrsum(divs) - i;
		if (divsum > i)
			abundants[len++] = i;
		free(divs);
	}
	//Determine which can be made up of sums of abundants
	for (int i = 0; abundants[i] + abundants[i+1] < 28123; i++) {
		for (int j = i; abundants[i] + abundants[j] < 28123; j++) {
			absums[abundants[i] + abundants[j]] = 0;
		}
	}
	//Turn the array of bools into a total
	for (int i = 1; i < 28123; i++) {
		if (absums[i])
			total += i;
	}
	printf("%lu\n", total);
	return 0;
}
